src/transformers/modeling_rag.py

```diff
@@ -0,0 +1,950 @@
+
+        outputs = self.generator(
+            input_ids,
+            attention_mask=attention_mask,
+            encoder_outputs=encoder_outputs,
+            decoder_input_ids=decoder_input_ids,
+            decoder_past_key_values=decoder_past_key_values,
+            use_cache=use_cache,
+            **generator_kwargs,
+        )
+        return Seq2SeqLMOutputWithDocs(
+            loss=None,
+            logits=outputs.logits,
+            decoder_past_key_values=outputs.decoder_past_key_values,
+            decoder_hidden_states=outputs.decoder_hidden_states,
+            decoder_attentions=outputs.decoder_attentions,
+            encoder_last_hidden_state=outputs.encoder_last_hidden_state,
+            encoder_hidden_states=outputs.encoder_hidden_states,
+            encoder_attentions=outputs.encoder_attentions,
+            doc_scores=doc_scores,
+        )
+
+    def _cat_input_and_doc(self, doc_score, doc_title, doc_text, input_string, add_eos, print_docs=False):
+        # TODO(Patrick): if we train more RAG models, I want to put the input first to take advantage of effortless truncation
+        # TODO(piktus): handle truncation
+        if doc_title.startswith('"'):
+            doc_title = doc_title[1:]
+        if doc_title.endswith('"'):
+            doc_title = doc_title[:-1]
+        prefix = self.generator.config.prefix if self.generator.config.prefix is not None else ""
+        suffix = self.generator_tokenizer.eos_token if add_eos else ""
+        out = (
+            prefix + doc_title + self.config.title_sep + doc_text + self.config.doc_sep + input_string + suffix
+        ).replace("  ", " ")
+        if print_docs:
+            logger.info("{} {}".format(doc_score, out))
+        return out
+
+    def contextualize(self, input_ids, print_docs=False):
+        r"""
+        Retrieves context documents for every input in the batch.
+
+        Args:
+            input_ids (:obj:`torch.LongTensor` of shape :obj:`(batch_size, sequence_length)`):
+                Indices of input sequence tokens in the vocabulary.
+            print_docs  (:obj:`bool`, `optional`, defaults to :obj:`True`):
+                If :obj:`True`, documents retrieved during the forward pass will be printed out. Intended for debugging purposes.
+
+        Return:
+            :obj:`tuple(tuple(torch.FloatTensor)`: a tuple consisting od three elements: contextualized ``input_ids``,
+                compatible ``attention_mask`` and scores of the retrieved documents.
+        """
+
+        device = input_ids.device
+        # T5 tokenizer doesn't add eos token by default evan with add_special_tokens set to True
+        add_eos = (input_ids == self.generator_tokenizer.eos_token_id).any() and isinstance(
+            self.generator, T5ForConditionalGeneration
+        )
+
+        input_strings = self.generator_tokenizer.batch_decode(input_ids, skip_special_tokens=True)
```

Comment:
Maybe we could create a new type of "composite" tokenizer which would provide inputs encoded using two different tokenizers?

Reply:
@patrickvonplaten do you need something like this for seq2seq models as well?